### PR TITLE
Update `c2rust-transpile` to use `clap` 3's `derive` API

### LINKED ISSUE #395: Update to clap3
Clap3 deprecated YAML support, so we need to rewrite command-line parsing to use the new APIs.

### PR DESCRIPTION
**Pull request by ivyraine**
- Fixes #395.

- Depends on #703 (branched off of it).

Previously, we used `transpile.yaml` to declare `clap` options. This PR aims to remove `transpile.yaml` as it's deprecated in clap 3+, and uses the more idiomatic `derive` args provided by clap.

It also adds in the missing `--fail-on-multiple` flag, which is already implemented but lacks a flag.

The `diff` between the using `cargo run -- transpile --help` before and after this PR (with before being #703) is as follows:

```diff
--- #703	2023-01-08 20:56:21.000000000 -0500
+++ #777	2023-01-13 21:16:25.000000000 -0500
@@ -1,19 +1,19 @@
-transpile 0.9.0
+transpile 0.16.0
 - The C2Rust Project Developers <c2rust@immunant.com>
 - Eric Mertens <emertens@galois.com>
 - Alec Theriault <atheriault@galois.com>
 Translate C code to equivalent Rust code

 USAGE:
-    c2rust-transpile [OPTIONS] <COMPILE_COMMANDS> [--] [extra-clang-args]...
+    c2rust-transpile [OPTIONS] <COMPILE_COMMANDS> [--] [EXTRA_CLANG_ARGS]...

 ARGS:
     <COMPILE_COMMANDS>       Input compile_commands.json file
-    <extra-clang-args>...    Extra arguments to pass to clang frontend during parsing the input
+    <EXTRA_CLANG_ARGS>...    Extra arguments to pass to clang frontend during parsing the input
                              C file

 OPTIONS:
-    -b, --binary <binary>
+    -b, --binary <BINARY>
             Emit Rust build files for a binary using the main function in the specified translation
             unit (implies -e/--emit-build-files)

@@ -43,7 +43,7 @@

     -e, --emit-build-files
             Emit Rust build files, i.e., Cargo.toml for a library (and one or more binaries if
-            -b/--binary is given). Implies --emit-modules.
+            -b/--binary is given). Implies --emit-modules

         --emit-modules
             Emit .rs files as modules instead of crates, excluding the crate preambles
@@ -51,12 +51,15 @@
         --emit-no-std
             Emit code using core rather than std

-    -f, --filter <filter>
+    -f, --filter <FILTER>
             Only transpile files matching filter

         --fail-on-error
             Fail to translate a module when a portion is not able to be translated

+        --fail-on-multiple
+            Fail when the control-flow graph generates branching constructs
+
     -h, --help
             Print help information

@@ -66,14 +69,14 @@
         --ignore-c-multiple-info
             Don't keep/use information about C branches

-        --invalid-code <invalid-code>
+        --invalid-code <INVALID_CODE>
             How to handle violated invariants or invalid code [default: compile_error] [possible
             values: panic, compile_error]

         --json-function-cfgs
             Dumps into files JSON visualizations of the CFGs of every function

-        --log-level <log-level>
+        --log-level <LOG_LEVEL>
             Logging level [default: warn] [possible values: off, error, warn, info, debug, trace]

         --no-incremental-relooper
@@ -84,12 +87,12 @@

     -o, --output-dir <DIR>
             Path to output directory. Rust sources will be emitted in DIR/src/ and build files will
-            be emitted in DIR/.
+            be emitted in DIR/

         --overwrite-existing
             Emit files even if it causes existing files to be overwritten

-        --prefix-function-names <prefix-function-names>
+        --prefix-function-names <PREFIX_FUNCTION_NAMES>
             Adds a prefix to all function names. Generally only useful for testing

         --preserve-unused-functions
@@ -109,7 +112,7 @@

         --translate-fn-macros
             Enable translation of some C function macros into invalid Rust code. WARNING: resulting
-            code will not compile.
+            code will not compile

     -v, --verbose
             Verbose mode
@@ -117,5 +120,5 @@
     -V, --version
             Print version information

-    -W <warn>
+    -W <WARN>
             Enable the specified warning (all enables all warnings)
```

These differences are expected and should have no functional differences:
- `clap` now automatically removes trailing periods from the doc
- `clap` now automatically capitalizes the user-input strings, as well as uses underscores instead of hyphens
- the correction `c2rust-transpile` version is now used (it was previously missed in releases)

`--fail-on-multiple` was also present in the `clap` Rust code, but not in the YAML, so that's been re-added.

Some changes need further testing:
- Need to check if the `EXTRA_CLANG_ARGS` is actually optional, as intended.
- Need to check that check that the options actually work. Does CI check these?
- Need to verify that the new `--fail-on-multiple` flag works as intended.

### CODE REVIEW

**Comment by ivyraine** [comment]
The `diff` between the using `cargo run -- transpile -h` before and after this PR is as follows:
```
1c1
< transpile 0.9.0
---
> transpile 0.16.0
8c8
<     c2rust-transpile [OPTIONS] <COMPILE_COMMANDS> [--] [extra-clang-args]...
---
>     c2rust-transpile [OPTIONS] <COMPILE_COMMANDS> [--] [EXTRA_CLANG_ARGS]...
12c12
<     <extra-clang-args>...    Extra arguments to pass to clang frontend during parsing the input
---
>     <EXTRA_CLANG_ARGS>...    Extra arguments to pass to clang frontend during parsing the input
16c16
<     -b, --binary <binary>
---
>     -b, --binary <BINARY>
46c46
<             -b/--binary is given). Implies --emit-modules.
---
>             -b/--binary is given). Implies --emit-modules
54c54
<     -f, --filter <filter>
---
>     -f, --filter <FILTER>
59a60,62
>         --fail-on-multiple
>             Fail when the control-flow graph generates branching constructs
> 
69,71c72,74
<         --invalid-code <invalid-code>
<             How to handle violated invariants or invalid code [default: compile_error] [possible
<             values: panic, compile_error]
---
>         --invalid-code <INVALID_CODE>
>             How to handle violated invariants or invalid code [default: compile-error] [possible
>             values: panic, compile-error]
76c79
<         --log-level <log-level>
---
>         --log-level <LOG_LEVEL>
87c90
<             be emitted in DIR/.
---
>             be emitted in DIR/
92c95
<         --prefix-function-names <prefix-function-names>
---
>         --prefix-function-names <PREFIX_FUNCTION_NAMES>
112c115
<             code will not compile.
---
>             code will not compile
120c123
<     -W <warn>
---
>     -W <WARN>
```
These differences are expected and should have no functional differences:
- `clap` now automatically removes trailing periods from the doc
- `clap` now automatically capitalizes the user-input strings, as well as uses underscores instead of hyphens

There is one potential exception in that the default `--invalid-code` is now `compile-error` rather than `compile_error`. This naming is generated by `clap` based on an enum field. From full repo searching, it appears that this option is not used in any scripts, so it is not a breaking change.

Some changes need further testing:
- Need to check if the `EXTRA_CLANG_ARGS` is actually optional, as intended.
- Need to check that check that the options actually work. Does CI check these?
- Need to verify that the new `--fail-on-multiple` flag works as intended.

File: c2rust/src/bin/c2rust-transpile.rs [hunk 1]
```diff
@@ -1,68 +1,234 @@
...
         filter: {
-            if matches.is_present("filter") {
-                let filter = matches.value_of("filter").unwrap();
-                Some(Regex::new(filter).unwrap())
+            if args.filter.is_some() {
+                let filter = args.filter.unwrap();
+                Some(Regex::new(filter.as_str()).unwrap())
             } else {
                 None
             }
```
Comment (kkysen) [inline c2rust/src/bin/c2rust-transpile.rs:228]:
```suggestion
            args.filter.map(|filter| Regex::new(filter.as_str()).unwrap())
```

File: c2rust/src/bin/c2rust-transpile.rs [hunk 2]
```diff
@@ -1,68 +1,234 @@
...
-    let extra_args: Vec<&str> = match matches.values_of("extra-clang-args") {
-        Some(args) => args.collect(),
-        None => Vec::new(),
-    };
 
-    let enabled_warnings: HashSet<Diagnostic> = matches
-        .values_of("warn")
-        .unwrap_or_default()
+    let extra_args = args.extra_clang_args;
+    let extra_args = extra_args.iter().map(AsRef::as_ref).collect::<Vec<&str>>();
```
Comment (kkysen) [inline c2rust/src/bin/c2rust-transpile.rs:190]:
```suggestion
    let extra_args = args.extra_clang_args.iter().map(AsRef::as_ref).collect::<Vec<&str>>();
```

File: c2rust/src/bin/c2rust-transpile.rs [hunk 3]
```diff
@@ -1,68 +1,234 @@
...
+    let extra_args = extra_args.as_slice();
+
+    let enabled_warnings: HashSet<Diagnostic> = args
+        .warn
+        .iter()
```
Comment (kkysen) [inline c2rust/src/bin/c2rust-transpile.rs:195]: I think `warn` was meant to be a `multiple` arg, but it wasn't, so this should be fine now.

File: c2rust/src/bin/c2rust-transpile.rs [hunk 4]
```diff
@@ -1,68 +1,234 @@
...
         .map(|s| Diagnostic::from_str(s).unwrap())
```
Comment (kkysen) [inline c2rust/src/bin/c2rust-transpile.rs:196]: Can `warn` be an `Option<Diagnostic>` using its `from_str`?

**Comment by kkysen** [comment]
> There is one potential exception in that the default `--invalid-code` is now `compile-error` rather than `compile_error`. This naming is generated by `clap` based on an enum field. From full repo searching, it appears that this option is not used in any scripts, so it is not a breaking change.

Fixed in 2245887aa67a2c5c8b80335c2b60ac4aaed364be.

File: c2rust/src/bin/c2rust-transpile.rs (hunk 4 above)
Comment (kkysen) [reply c2rust/src/bin/c2rust-transpile.rs]: Fixed in 5c6d4de333fe7b1cdb592c9f58169ba39e3e4bed.

**Review by kkysen — APPROVED** [review]

**Review by fw-immunant — APPROVED** [review]
LGTM. Thanks for the cleanup!

**Pull request merged**